File: AtividadesPosixJava/LeitoresEscritoresSemaforo.java
import java.util.concurrent.Semaphore;

public class LeitoresEscritoresSemaforo {

    static int sharedData = 0;
    static int readCount = 0;
    static int writeCount = 0;

    static Semaphore mutexReadCount = new Semaphore(1);
    static Semaphore mutexWriteCount = new Semaphore(1);
    static Semaphore resourceAccess = new Semaphore(1);
    static Semaphore readTry = new Semaphore(1);

    static class Escritor implements Runnable {
        private final int id;
        public Escritor(int id) { this.id = id; }
        
        @Override
        public void run() {
            try {
                while(true) {
                    mutexWriteCount.acquire();
                    writeCount++;
                    if (writeCount == 1) readTry.acquire();
                    mutexWriteCount.release();

                    resourceAccess.acquire();
                    sharedData++;
                    System.out.printf("Escritor %d escreveu: %d\n", id, sharedData);
                    resourceAccess.release();

                    mutexWriteCount.acquire();
                    writeCount--;
                    if (writeCount == 0) readTry.release();
                    mutexWriteCount.release();

                    Thread.sleep((long) (Math.random() * 5000 + 2000));
                }
            } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        }
    }

    static class Leitor implements Runnable {
        private final int id;
        public Leitor(int id) { this.id = id; }

        @Override
        public void run() {
            try {
                while(true) {
                    readTry.acquire();
                    mutexReadCount.acquire();
                    readCount++;
                    if (readCount == 1) resourceAccess.acquire();
                    mutexReadCount.release();
                    readTry.release();

                    System.out.printf("Leitor %d leu: %d (leitores ativos: %d)\n", id, sharedData, readCount);

                    mutexReadCount.acquire();
                    readCount--;
                    if (readCount == 0) resourceAccess.release();
                    mutexReadCount.release();

                    Thread.sleep((long) (Math.random() * 3000 + 1000));
                }
            } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        }
    }

    public static void main(String[] args) {
        for (int i = 1; i <= 2; i++) new Thread(new Escritor(i)).start();
        for (int i = 1; i <= 5; i++) new Thread(new Leitor(i)).start();
    }
}

File: AtividadesPosix/leitores_escritores_semaforo.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

int shared_data = 0;
int read_count = 0;
int write_count = 0;

sem_t mutex_read_count;
sem_t mutex_write_count;
sem_t resource_access;
sem_t read_try;

void* escritor(void* arg) {
    int id = *(int*)arg;
    free(arg);

    while(1) {
        sleep(rand() % 5 + 2);
        
        sem_wait(&mutex_write_count);
        write_count++;
        if (write_count == 1) {
            sem_wait(&read_try);
        }
        sem_post(&mutex_write_count);

        sem_wait(&resource_access);
        shared_data++;
        printf("Escritor %d escreveu: %d\n", id, shared_data);
        sem_post(&resource_access);

        sem_wait(&mutex_write_count);
        write_count--;
        if (write_count == 0) {
            sem_post(&read_try);
        }
        sem_post(&mutex_write_count);
    }
    return NULL;
}

void* leitor(void* arg) {
    int id = *(int*)arg;
    free(arg);

    while(1) {
        sleep(rand() % 3 + 1);

        sem_wait(&read_try);
        sem_wait(&mutex_read_count);
        read_count++;
        if (read_count == 1) {
            sem_wait(&resource_access);
        }
        sem_post(&mutex_read_count);
        sem_post(&read_try);

        printf("Leitor %d leu: %d (leitores ativos: %d)\n", id, shared_data, read_count);

        sem_wait(&mutex_read_count);
        read_count--;
        if (read_count == 0) {
            sem_post(&resource_access);
        }
        sem_post(&mutex_read_count);
    }
    return NULL;
}

int main() {
    pthread_t leitores[5], escritores[2];
    
    sem_init(&mutex_read_count, 0, 1);
    sem_init(&mutex_write_count, 0, 1);
    sem_init(&resource_access, 0, 1);
    sem_init(&read_try, 0, 1);

    for(int i = 0; i < 2; i++) { int* id = malloc(sizeof(int)); *id = i + 1; pthread_create(&escritores[i], NULL, escritor, id); }
    for(int i = 0; i < 5; i++) { int* id = malloc(sizeof(int)); *id = i + 1; pthread_create(&leitores[i], NULL, leitor, id); }

    for(int i = 0; i < 2; i++) pthread_join(escritores[i], NULL);
    for(int i = 0; i < 5; i++) pthread_join(leitores[i], NULL);

    sem_destroy(&mutex_read_count);
    sem_destroy(&mutex_write_count);
    sem_destroy(&resource_access);
    sem_destroy(&read_try);

    return 0;
}